src/mlio-py/mlio/core/data_reader.cxx

```diff
@@ -142,86 +166,55 @@ make_csv_reader(
-    rdr_prm.batch_size = batch_size;
-    rdr_prm.num_prefetched_batches = num_prefetched_batches;
-    rdr_prm.num_parallel_reads = num_parallel_reads;
-    rdr_prm.last_batch_hnd = last_batch_hnd;
-    rdr_prm.bad_batch_hnd = bad_batch_hnd;
-    rdr_prm.num_instances_to_skip = num_instances_to_skip;
-    rdr_prm.num_instances_to_read = num_instances_to_read;
-    rdr_prm.shard_index = shard_index;
-    rdr_prm.num_shards = num_shards;
-    rdr_prm.shuffle_instances = shuffle_instances;
-    rdr_prm.shuffle_window = shuffle_window;
-    rdr_prm.shuffle_seed = shuffle_seed;
-    rdr_prm.reshuffle_each_epoch = reshuffle_each_epoch;
-    rdr_prm.subsample_ratio = subsample_ratio;
-
-    csv_params csv_prm{};
-
-    csv_prm.column_names = std::move(column_names);
-    csv_prm.name_prefix = std::move(name_prefix);
-    csv_prm.use_columns = std::move(use_columns);
-    csv_prm.use_columns_by_index = std::move(use_columns_by_index);
-    csv_prm.default_data_type = default_data_type;
-    csv_prm.column_types = std::move(column_types);
-    csv_prm.column_types_by_index = std::move(column_types_by_index);
-    csv_prm.header_row_index = header_row_index;
-    csv_prm.has_single_header = has_single_header;
-    csv_prm.delimiter = delimiter;
-    csv_prm.quote_char = quote_char;
-    csv_prm.comment_char = comment_char;
-    csv_prm.allow_quoted_new_lines = allow_quoted_new_lines;
-    csv_prm.skip_blank_lines = skip_blank_lines;
+    std::shared_ptr<csv_params> csv_prm = std::make_shared<csv_params>();
+
+    csv_prm->column_names = std::move(column_names);
+    csv_prm->name_prefix = std::move(name_prefix);
+    csv_prm->use_columns = std::move(use_columns);
+    csv_prm->use_columns_by_index = std::move(use_columns_by_index);
+    csv_prm->default_data_type = default_data_type;
+    csv_prm->column_types = std::move(column_types);
+    csv_prm->column_types_by_index = std::move(column_types_by_index);
+    csv_prm->header_row_index = header_row_index;
+    csv_prm->has_single_header = has_single_header;
+    csv_prm->delimiter = delimiter;
+    csv_prm->quote_char = quote_char;
+    csv_prm->comment_char = comment_char;
+    csv_prm->allow_quoted_new_lines = allow_quoted_new_lines;
+    csv_prm->skip_blank_lines = skip_blank_lines;
     if (!encoding.empty()) {
-        csv_prm.encoding = text_encoding{std::move(encoding)};
+        csv_prm->encoding = text_encoding{std::move(encoding)};
+    }
+    csv_prm->max_field_length = max_field_length;
+    csv_prm->max_field_length_hnd = max_field_length_hnd;
+    csv_prm->max_line_length = max_line_length;
+    csv_prm->parser_prm.nan_values = std::move(nan_values);
+    csv_prm->parser_prm.base = base;
+
+    return csv_prm;
+}
+
+intrusive_ptr<csv_reader>
+make_csv_reader(std::shared_ptr<data_reader_params> rdr_prm_ptr,
+                std::optional<std::shared_ptr<csv_params>> csv_prm_ptr)
+{
+    // Copy construct parameters
+    data_reader_params rdr_prm(*rdr_prm_ptr);
```

Comment:
You won't need this line if you take your arguments by value, but just as a side note, use brace initialization syntax like `rdr_prm{*rdr_prm_ptr}` to be consistent with the rest of the codebase.

Reply:
Does that also perform copy construction? I wasn't sure, so I went with what I found in the [cpp reference](https://en.cppreference.com/w/cpp/language/copy_constructor)